src/canister_tests/src/tests.rs

```diff
@@ -955,6 +956,262 @@ mod http_tests {
+        env.advance_time(Duration::from_secs(365 * 24 * 60 * 60));
+        // preparing a new delegation prunes old ones
+        api::prepare_delegation(
+            &env,
+            canister_id,
+            principal_1(),
+            user_number,
+            frontend_hostname.to_string(),
+            ByteBuf::from("last session key"),
+            None,
+        )?;
+
+        let metrics = flows::get_metrics(&env, canister_id);
+        let (signature_count, _) =
+            framework::parse_metric(&metrics, "internet_identity_signature_count");
+        assert_eq!(signature_count, 1); // old ones pruned and a new one created
+        Ok(())
+    }
+
+    /// Verifies that the stable memory pages count metric is updated correctly.
+    #[test]
+    fn metrics_stable_memory_pages_should_increase_with_more_users() -> Result<(), CallError> {
+        let env = StateMachine::new();
+        let canister_id = framework::install_ii_canister(&env, framework::II_WASM.clone());
+
+        let metrics = flows::get_metrics(&env, canister_id);
+        let (signature_count, _) =
+            framework::parse_metric(&metrics, "internet_identity_stable_memory_pages");
+        // empty II has some metadata in stable memory which requires at least one page
+        assert_eq!(signature_count, 1);
+
+        // a wasm page is 64kb and a single user takes up 2kb -> 32 users require a complete wasm page
+        for _ in 0..32 {
+            flows::register_anchor(&env, canister_id);
+        }
+
+        let metrics = flows::get_metrics(&env, canister_id);
+        let (signature_count, _) =
+            framework::parse_metric(&metrics, "internet_identity_stable_memory_pages");
+        assert_eq!(signature_count, 2);
+        Ok(())
+    }
+
+    /// Verifies that the last II wasm upgrade timestamp is updated correctly.
+    #[test]
+    fn metrics_last_upgrade_timestamp_should_update_after_upgrade() -> Result<(), CallError> {
+        let env = StateMachine::new();
+        let canister_id = framework::install_ii_canister(&env, framework::II_WASM.clone());
+        // immediately upgrade because installing the canister does not set the metric
+        framework::upgrade_ii_canister(&env, canister_id, framework::II_WASM.clone());
+
+        for _ in 0..2 {
+            let metrics = flows::get_metrics(&env, canister_id);
+            let (upgrade_timestamp, _) =
+                framework::parse_metric(&metrics, "internet_identity_last_upgrade_timestamp");
+            assert_eq!(
+                upgrade_timestamp,
+                env.time()
+                    .duration_since(SystemTime::UNIX_EPOCH)
+                    .unwrap()
+                    .as_nanos() as u64
+            );
+
+            env.advance_time(Duration::from_secs(300)); // the state machine does not advance time on its own
+            framework::upgrade_ii_canister(&env, canister_id, framework::II_WASM.clone());
```

Comment:
here again I'm a bit confused by the construct of "first check metric, then perform operation that we expect to change the metrics". Do you mean to check the metrics after the op? if not, what does upgrading the canister give you on the last loop?

Reply:
I wanted to avoid having two assert metrics, so I figured I could just loop (and disregard the last loop execution). But this is confusing and for this test I have now removed the loop altogether.